linkerd/app/outbound/src/tcp/connect.rs

```diff
@@ -1,77 +1,107 @@
 use super::opaque_transport::OpaqueTransport;
-use crate::target::Endpoint;
+use crate::{target::Endpoint, Outbound};
 use linkerd_app_core::{
-    config::ConnectConfig,
-    io, metrics,
-    proxy::identity::LocalCrtKey,
-    svc, tls,
-    transport::{ConnectAddr, ConnectTcp},
-    transport_header::SessionProtocol,
-    Error,
+    io, svc, tls, transport::ConnectTcp, transport_header::SessionProtocol, Error, ProxyRuntime,
 };
 use tracing::debug_span;
 
-// Establishes connections to remote peers (for both TCP forwarding and HTTP
-// proxying).
-pub fn stack<P>(
-    config: &ConnectConfig,
-    server_port: u16,
-    local_identity: Option<LocalCrtKey>,
-    metrics: &metrics::Proxy,
-) -> impl svc::Service<
-    Endpoint<P>,
-    Response = impl tokio::io::AsyncRead + tokio::io::AsyncWrite + Send + Unpin,
-    Error = Error,
-    Future = impl Send,
-> + Clone
-where
-    Endpoint<P>: svc::stack::Param<Option<SessionProtocol>> + svc::stack::Param<ConnectAddr>,
-{
-    let identity_disabled = local_identity.is_none();
-    svc::stack(ConnectTcp::new(config.keepalive))
-        // Initiates mTLS if the target is configured with identity. The
-        // endpoint configures ALPN when there is an opaque transport hint OR
-        // when an authority override is present (indicating the target is a
-        // remote cluster gateway).
-        .push(tls::Client::layer(local_identity))
-        // Encodes a transport header if the established connection is TLS'd and
-        // ALPN negotiation indicates support.
-        .push(OpaqueTransport::layer())
-        // Limits the time we wait for a connection to be established.
-        .push_timeout(config.timeout)
-        .push(svc::stack::BoxFuture::layer())
-        .push(metrics.transport.layer_connect())
-        .push_map_target(move |e: Endpoint<P>| {
-            if identity_disabled {
-                e.identity_disabled()
-            } else {
-                e
-            }
-        })
-        .push_request_filter(PreventLoop { port: server_port })
-        .into_inner()
+pub fn stack(config: crate::Config, runtime: ProxyRuntime) -> Outbound<ConnectTcp> {
+    let stack = svc::stack(ConnectTcp::new(config.proxy.connect.keepalive));
+    Outbound {
+        config,
+        runtime,
+        stack,
+    }
```

Comment:
nit/TIOLI: seems like this could alternatively be a function on the `Outbound` type, like `Outbound::new_tcp_connect`? not sure if that's better or worse.